### PR TITLE
*: add backup writer

### PR DESCRIPTION
**Pull request by fanminshi**
expose parseBucketAndKey so that writer can also parse s3 path into bucket and key.
add writer interface/impl.

### CODE REVIEW

**Comment by fanminshi** [comment]
cc/ @hongchaodeng

File: pkg/backup/writer/s3_writer.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+		return err
+	}
+
+	_, err = s3w.u.Upload(
+		&s3manager.UploadInput{
+			Bucket: aws.String(bk),
+			Key:    aws.String(key),
+			Body:   r,
+		})
+	if err != nil {
```
Comment (hongchaodeng) [inline pkg/backup/writer/s3_writer.go]: return err

Comment (fanminshi) [reply pkg/backup/writer/s3_writer.go]: done

**Comment by fanminshi** [comment]
merge when green.

**Comment by hasbro17** [comment]
LGTM

**Comment by fanminshi** [comment]
via https://jenkins-etcd-public.prod.coreos.systems/job/etcd-operator/2760/console

```18:38:48 --- FAIL: TestPauseControl (60.19s)
18:38:48 	crd_util.go:45: creating etcd cluster: test-etcd-tphlp
18:38:48 	util.go:82: 2017-10-30 18:35:19.247636839 +0000 UTC m=+158.444391148 waiting size (3), healthy etcd members: names ([])
18:38:48 	util.go:82: 2017-10-30 18:35:29.247960095 +0000 UTC m=+168.444714437 waiting size (3), healthy etcd members: names ([])
18:38:48 	util.go:82: 2017-10-30 18:35:39.247830536 +0000 UTC m=+178.444584878 waiting size (3), healthy etcd members: names ([test-etcd-tphlp-0000])
18:38:48 	util.go:82: 2017-10-30 18:35:49.247916847 +0000 UTC m=+188.444671216 waiting size (3), healthy etcd members: names ([test-etcd-tphlp-0000])
18:38:48 	util.go:82: 2017-10-30 18:35:59.247883342 +0000 UTC m=+198.444637655 waiting size (3), healthy etcd members: names ([test-etcd-tphlp-0000])
18:38:48 	util.go:82: 2017-10-30 18:36:09.249107089 +0000 UTC m=+208.445861442 waiting size (3), healthy etcd members: names ([test-etcd-tphlp-0000])
18:38:48 	util.go:82: 2017-10-30 18:36:19.247913544 +0000 UTC m=+218.444667841 waiting size (3), healthy etcd members: names ([test-etcd-tphlp-0000])
18:38:48 	basic_test.go:68: failed to create 3 members etcd cluster: still failing after 6 retries
18:38:48 	crd_util.go:77: deleting etcd cluster: test-etcd-tphlp
18:38:48 FAIL
```

**Comment by fanminshi** [comment]
@etcd-bot re-test this please reason: jenkins-ci flake.

**Pull request merged**